gui/masspit.lua

```diff
@@ -0,0 +1,218 @@
+-- Interface to assign caged units in a stockpile to pond/pit.
+
+local utils = require 'utils'
+local gui = require('gui')
+local widgets = require('gui.widgets')
+
+Masspit = defclass(Masspit, widgets.Window)
+Masspit.ATTRS {
+    frame_title='masspit',
+    frame={w=43, h=22},
+    resizable=true,
+    resize_min={w=43, h=12},
+}
+
+function Masspit:init()
+    self:addviews{
+        widgets.WrappedLabel{
+            view_id = 'label',
+            frame = { l = 0, t = 0 },
+            text_to_wrap = "Select or click on a stockpile"
+        },
+        widgets.Pages{
+            view_id ='pages',
+            subviews = {
+                widgets.List{
+                    view_id = 'stockpiles',
+                    frame = { l = 0, t = 4 },
+                    on_select = self:callback("focusOnZone"),
+                    on_submit = self:callback("setStockpile")
+                },
+                widgets.List{
+                    view_id = 'pits',
+                    frame = { l = 0, t = 3 },
+                    on_select = self:callback("focusOnZone"),
+                    on_submit = self:callback("setPit")
+                },
+                widgets.List{
+                    view_id = 'pitted',
+                    frame = { l = 0, t = 2 },
+                },
+            }
+        },
+    }
+
+    self:showStockpiles()
+end
+
+function Masspit:showStockpiles()
+    self.subviews.pages:setSelected('stockpiles')
+
+    local choices = {}
+
+    for _, zone in pairs(df.global.world.buildings.other.STOCKPILE) do
+        if (#zone.settings.animals.enabled > 0) then
+            local zone_name = zone.name.length and dfhack.TranslateName(zone.name) or "Animal stockpile #" .. zone.stockpile_number
+            local zone_position = df.coord:new()
+            zone_position.x = zone.centerx
+            zone_position.y = zone.centery
+            zone_position.z = zone.z
+
+            table.insert(choices, {
```

Comment:
how about checking these stockpiles for caged creatures *before* adding them to the choices list. that way the player can only select valid choices

Reply:
Yeah this is a good point, I thought the iteration over stockpile items might cause performance issues in larger forts though so I wasn't sure. What do you think, since we are using `getStockpileContents` and trying to find any caged creature this could cause issues for very large stockpiles?